[docs] Query tuning intro (#8582)
yugabyte/yugabyte-db: docs/content/latest/explore/query-1-performance/query-tuning-intro.md

status        | character varying(25)       |           |          |
first_name    | character varying(55)       |           |          |
last_name     | character varying(55)       |           |          |
phone         | character varying(25)       |           |          |
created_at    | timestamp without time zone |           | not null |
updated_at    | timestamp without time zone |           | not null |
product_id    | integer                     |           |          |
business_type | character varying(55)       |           |          |
sales_rep     | character varying           |           |          |

Indexes:
    "accounts_pkey" PRIMARY KEY, lsm (id HASH)
    "index_on_company_name" lsm (company_name HASH)
```

`EXPLAIN` output for the query is as follows:

```sql
yugabyte=# explain SELECT accounts.* FROM accounts 
    ORDER BY accounts.id desc, accounts.id desc LIMIT 25 OFFSET 0;
```

```output
QUERY PLAN
----------------------------------------------
Limit (cost=128.22..128.28 rows=25 width=1642)
    Sort (cost=128.22..130.72 rows=1000 width=1642)
        Sort Key: id DESC
            Seq Scan on accounts (cost=0.00..100.00 rows=1000 width=1642)
```

> In this case, a sort is run first, which adds extra time to the query, before running a sequential scan of the table, which, as we have seen, also degrades performance.

Reviewer: One missed "we", gonna just commit this for you :) 

```suggestion
In this case, a sort is run first, which adds extra time to the query, before running a sequential scan of the table, which also degrades performance.
```

---
title: Introduction to Query Tuning
linkTitle: Introduction
description: Introduction to Query Tuning
aliases:
headerTitle: Introduction to Query Tuning
image: /images/section_icons/index/develop.png
menu:
  latest:
    identifier: query-tuning-intro
    parent: query-tuning
    weight: 150
isTocNested: true
showAsideToc: true
---

> Improving the performance of queries in YugabyteDB typically involves some combination of the following general steps:

Author: fixed

**1. Find slow queries using pg_stat_statements** 

Use the `pg_stat_statements` extension get statistics on the execution of queries. Using `pg_stat_statements`, you can quickly track down problematic queries by a variety of criteria, including the following:

* I/O intensity
* time consumption
* response time
* memory consumption
* temporary space consumption

The [`pg_stat_statements`](https://www.postgresql.org/docs/11/pgstatstatements.html) extension module is installed by default, but must be enabled before the `pg_stat_statements` view can be queried.

```sql

> CREATE EXTENSION pg_stat_statements;

Author: Fixed

    Seq Scan on contacts (cost=0.00..105.00 rows=1000 width=0)
        Filter: (is_paid AND is_over_charged AND (updated_at > '2021-04-12 12:00:00'::timestamp without time zone) AND (account_id = 5678))
(3 rows)
```

```sql
yugabyte=# explain SELECT COUNT(*) FROM contacts WHERE contacts.account_id = 7890 
    AND contacts.is_paid = FALSE 
    AND (updated_at > '2021-04-12 12:00:00');
```

```output
QUERY PLAN
----------------------------------------------
Aggregate (cost=107.50..107.51 rows=1 width=8)
    Seq Scan on contacts (cost=0.00..105.00 rows=1000 width=0)
        Filter: ((NOT is_paid) AND (updated_at > '2021-04-12 12:00:00'::timestamp without time zone) AND (account_id =1234))
(3 rows)
```

In each case, the queries do a `Seq Scan` (sequential scan) on the tables. This operation requires scanning the entire table to retrieve the desired columns. Even though we are using the partition keys to do the look up, it still needs to do a lot of scanning. 

Generally, you want to avoid `SELECT COUNT(*)` queries as they can require a full scan of the table to get the results. This can cause query degradation and in some cases cause the query to not return at all.

> Because most of the queries above use `account_id` as the main qualifier, we can avaoid a sequential scan by creating a direct index on that column, and then use the `INCLUDE` feature to cover the other columns that we also want in the index. Indexing is a powerful tool that can speed up queries with higher latencies. When creating an index, we want to consider the column cardinality, as well as the different index types.

Reviewer: ```suggestion
Because most of the queries above use `account_id` as the main qualifier, you can avoid a sequential scan by creating a direct index on that column, and then using the `INCLUDE` feature to cover the other columns that you also want in the index. Indexing is a powerful tool that can speed up queries with higher latencies. When creating an index, consider the column cardinality, as well as the different index types.
```

> We create the index as follows:

Reviewer: ```suggestion
Create the index as follows:
```

phone         | character varying(25)       |           |          |
created_at    | timestamp without time zone |           | not null |
updated_at    | timestamp without time zone |           | not null |
product_id    | integer                     |           |          |
business_type | character varying(55)       |           |          |
sales_rep     | character varying           |           |          |

Indexes:
    "accounts_pkey" PRIMARY KEY, lsm (id HASH)
    "index_on_company_name" lsm (company_name HASH)
```

`EXPLAIN` output for the query is as follows:

```sql
yugabyte=# explain SELECT accounts.* FROM accounts 
    ORDER BY accounts.id desc, accounts.id desc LIMIT 25 OFFSET 0;
```

```output
QUERY PLAN
----------------------------------------------
Limit (cost=128.22..128.28 rows=25 width=1642)
    Sort (cost=128.22..130.72 rows=1000 width=1642)
        Sort Key: id DESC
            Seq Scan on accounts (cost=0.00..100.00 rows=1000 width=1642)
```

In this case, a sort is run first, which adds extra time to the query, before running a sequential scan of the table, which, as we have seen, also degrades performance.

> To optimize this query, we can adjust the sorting of the table by the primary key to be `DESC` rather than `HASH`.

Reviewer: ```suggestion
To optimize this query, adjust the sorting of the table by the primary key to be `DESC` rather than `HASH`.
```

```sql
CREATE TABLE public.accounts ( id bigint NOT NULL, ... ,PRIMARY KEY(id desc);
```

> The result is the query no longer does a sequential scan; instead it uses an index scan, cutting execution time from 460ms to 3ms.

Reviewer: ```suggestion
Now, the query no longer does a sequential scan; instead it uses an index scan, cutting execution time from 430ms to 3ms.
```

```sql
yugabyte=# explain SELECT accounts.* FROM accounts 
    ORDER BY accounts.id desc, accounts.id desc LIMIT 25 OFFSET 0;
```

```output
QUERY PLAN
------------------------------------------
Limit (cost=0.00..2.85 rows=25 width=1642)
    Index Scan using accounts_pkey on accounts (cost=0.00..114.00 rows=1000 width=1642)

Time: 2.994 ms
Previous run time: 426.627 ms
```

> ### Optimizing `SELECT` using an index

Reviewer: All of these headings could be imperative. It's a decision we should make across the board sometime, but it works well in this doc.

```suggestion
### Optimize `SELECT` using an index
```

> The following example shows another query running a sequential scan, which we can fix by adding an index.

Reviewer: ```suggestion
The query in the following example runs a sequential scan, which you can fix by adding an index.
```

------------+--------------------------------+-----------+----------+--------
id          | bigint                         |           | not null | nextval('account_domains_id_seq'::regclass)
account_id  | integer                        |           | For      |
type        | character varying(55)          |           |          |
url         | character varying              |           |          |
is_valid    | boolean                        |           |          |
created_at  | timestamp(6) without time zone |           | not null |
updated_at  | timestamp(6) without time zone |           | not null |
verified_at | timestamp without time zone    |           |          |

Indexes:
    "account_type_pkey" PRIMARY KEY, lsm (id HASH)
```

```sql
yugabyte=# explain SELECT account_domains.* FROM account_type 
    WHERE account_type.account_id = 1234 ORDER BY account_type.url ASC;
```

```output
QUERY PLAN
----------------------------------------------
Sort (cost=152.33..154.83 rows=1000 width=237)
    Sort Key: url
        Seq Scan on account_type (cost=0.00..102.50 rows=1000 width=237)
            Filter: (account_id = 6873)
```

> The query runs a sequential scan on the account_type table, so by adding an index on `account_id`, we can prevent the full scan of the table, as follows:

Reviewer: ```suggestion
The query runs a sequential scan on the account_type table, so adding an index on `account_id` prevents the full scan of the table, as follows:
```
